S3: protocol detection fails when `ContentType` header set to MIME type of object data being uploaded

For example, CI run https://github.com/fsspec/s3fs/actions/runs/25286966929/job/74290822498

This is only seen on a few tests, but I am not sure what triggers it.

Full traceback in the client
<details>

```
func = <bound method ClientCreator._create_api_method.<locals>._api_call of <aiobotocore.client.S3 object at 0x7f1b7809d8d0>>

    async def _error_wrapper(func, *, args=(), kwargs=None, retries):
        if kwargs is None:
            kwargs = {}
        err = None
        for i in range(retries):
            wait_time = min(1.7**i * 0.1, 15)
    
            try:
>               return await func(*args, **kwargs)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^

s3fs/core.py:169: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiobotocore.client.S3 object at 0x7f1b7809d8d0>
                ]:
                    if pattern in str(e):
                        matched = True
                        break
    
                if matched:
                    await asyncio.sleep(wait_time)
                else:
                    should_retry = CUSTOM_ERROR_HANDLER(e)
                    if should_retry:
                        await asyncio.sleep(wait_time)
                    else:
                        break
            except Exception as e:
                err = e
                should_retry = CUSTOM_ERROR_HANDLER(e)
                if should_retry:
                    await asyncio.sleep(wait_time)
                else:
                    logger.debug("Nonretryable error: %s", e)
                    break
    
        if "'coroutine'" in str(err):
            # aiobotocore internal error - fetch original botocore error
            tb = err.__traceback__
            while tb.tb_next:
                tb = tb.tb_next
            try:
                await tb.tb_frame.f_locals["response"]
            except Exception as e:
                err = e
        err = translate_boto_error(err)
>       raise err
E       OSError: [Errno 121] Internal Server Error
```
</details>

Logging output from the server

<details>

```
INFO     werkzeug:_internal.py:97 127.0.0.1 - - [04/May/2026 19:29:27] "POST /test/temp?uploads HTTP/1.1" 500 -
ERROR    werkzeug:_internal.py:97 Error on request:
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 371, in run_wsgi
    execute(self.server.app)
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 332, in execute
    application_iter = app(environ, start_response)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/moto_server/werkzeug_app.py", line 316, in __call__
    return backend_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1536, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1514, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 919, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/utils.py", line 124, in __call__
    result = self.callback(request, request.url, dict(request.headers))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 368, in key_dispatch
    return cls().key_response(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1855, in key_response
    response = self._key_response(request, full_url)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1940, in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 3122, in _key_response_post
    status, resp_headers, resp_body = self.serialized(
                                      ^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 507, in serialized
    protocol = self.determine_response_protocol(service_model)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 495, in determine_response_protocol
    protocol = determine_request_protocol(service_model, content_type)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/request.py", line 50, in determine_request_protocol
    raise NotImplementedError(
NotImplementedError: Unsupported protocol [rest-json] for service s3
INFO     werkzeug:_internal.py:97 127.0.0.1 - - [04/May/2026 19:29:27] "POST /test/temp?uploads HTTP/1.1" 500 -
ERROR    werkzeug:_internal.py:97 Error on request:
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 371, in run_wsgi
    execute(self.server.app)
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 332, in execute
    application_iter = app(environ, start_response)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/moto_server/werkzeug_app.py", line 316, in __call__
    return backend_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1536, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1514, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 919, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/utils.py", line 124, in __call__
    result = self.callback(request, request.url, dict(request.headers))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 368, in key_dispatch
    return cls().key_response(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1855, in key_response
    response = self._key_response(request, full_url)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1940, in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 3122, in _key_response_post
    status, resp_headers, resp_body = self.serialized(
                                      ^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 507, in serialized
    protocol = self.determine_response_protocol(service_model)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 495, in determine_response_protocol
    protocol = determine_request_protocol(service_model, content_type)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/request.py", line 50, in determine_request_protocol
    raise NotImplementedError(
NotImplementedError: Unsupported protocol [rest-json] for service s3
INFO     werkzeug:_internal.py:97 127.0.0.1 - - [04/May/2026 19:29:28] "POST /test/temp?uploads HTTP/1.1" 500 -
ERROR    werkzeug:_internal.py:97 Error on request:
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 371, in run_wsgi
    execute(self.server.app)
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 332, in execute
    application_iter = app(environ, start_response)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/moto_server/werkzeug_app.py", line 316, in __call__
    return backend_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1536, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1514, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 919, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/utils.py", line 124, in __call__
    result = self.callback(request, request.url, dict(request.headers))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 368, in key_dispatch
    return cls().key_response(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1855, in key_response
    response = self._key_response(request, full_url)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1940, in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 3122, in _key_response_post
    status, resp_headers, resp_body = self.serialized(
                                      ^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 507, in serialized
    protocol = self.determine_response_protocol(service_model)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 495, in determine_response_protocol
    protocol = determine_request_protocol(service_model, content_type)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/request.py", line 50, in determine_request_protocol
    raise NotImplementedError(
NotImplementedError: Unsupported protocol [rest-json] for service s3
INFO     werkzeug:_internal.py:97 127.0.0.1 - - [04/May/2026 19:29:31] "POST /test/temp?uploads HTTP/1.1" 500 -
ERROR    werkzeug:_internal.py:97 Error on request:
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 371, in run_wsgi
    execute(self.server.app)
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/werkzeug/serving.py", line 332, in execute
    application_iter = app(environ, start_response)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/moto_server/werkzeug_app.py", line 316, in __call__
    return backend_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1536, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1514, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 919, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/utils.py", line 124, in __call__
    result = self.callback(request, request.url, dict(request.headers))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 368, in key_dispatch
    return cls().key_response(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1855, in key_response
    response = self._key_response(request, full_url)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 1940, in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/s3/responses.py", line 3122, in _key_response_post
    status, resp_headers, resp_body = self.serialized(
                                      ^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 507, in serialized
    protocol = self.determine_response_protocol(service_model)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/responses.py", line 495, in determine_response_protocol
    protocol = determine_request_protocol(service_model, content_type)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/miniconda/envs/test/lib/python3.11/site-packages/moto/core/request.py", line 50, in determine_request_protocol
    raise NotImplementedError(
NotImplementedError: Unsupported protocol [rest-json] for service s3
INFO     werkzeug:_internal.py:97 127.0.0.1 - - [04/May/2026 19:29:33] "POST /test/temp?uploads HTTP/1.1" 500 -
```
</details>

Seeing this as well, with `Unsupported protocol [rest-json] for service s3` (related to https://github.com/getmoto/moto/pull/9772?)

<details><summary>Full pytest traceback</summary>
<p>

```python
    @contextmanager
    def get_writer(self, path: str) -> Iterator[TextIOWrapper]:
        """Get writer for given path.
    
        Args:
            path: the path to get writer for.
    
        Yields:
            A TextIOWrapper to read the file/blob.
        """
        transport_params = {"client": self.client} if self.client else {}
        transport_params.update(self.extra_transport_params)
        try:
>           with smart_open.open(
                self.uri_with_path(path),
                "w+",
                transport_params=transport_params,
            ) as writer:

src/meltano/core/state_store/filesystem.py:144: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.nox/pytest-3-14/lib/python3.14/site-packages/smart_open/smart_open_lib.py:217: in open
    binary = _open_binary_stream(uri, binary_mode, transport_params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.nox/pytest-3-14/lib/python3.14/site-packages/smart_open/smart_open_lib.py:405: in _open_binary_stream
    fobj = submodule.open_uri(uri, mode, transport_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.nox/pytest-3-14/lib/python3.14/site-packages/smart_open/s3.py:310: in open_uri
    return open(parsed_uri['bucket_id'], parsed_uri['key_id'], mode, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

bucket_id = 'meltano', key_id = 'state/KFwtYBkdoMlYPucCPiBi/state.json', mode = 'wb+', version_id = None, buffer_size = 131072, min_part_size = 52428800
multipart_upload = True, defer_seek = False, client = <botocore.client.S3 object at 0xffb7e9880c20>
client_kwargs = {'S3.Client.create_multipart_upload': {'ContentType': 'application/json'}}, writebuffer = None, range_chunk_size = None
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.nox/pytest-3-14/lib/python3.14/site-packages/moto/s3/responses.py:1940: in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.nox/pytest-3-14/lib/python3.14/site-packages/moto/s3/responses.py:3122: in _key_response_post
    status, resp_headers, resp_body = self.serialized(
.nox/pytest-3-14/lib/python3.14/site-packages/moto/core/responses.py:507: in serialized
    protocol = self.determine_response_protocol(service_model)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.nox/pytest-3-14/lib/python3.14/site-packages/moto/core/responses.py:495: in determine_response_protocol
    protocol = determine_request_protocol(service_model, content_type)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

service_model = ServiceModel(s3), content_type = 'application/json'

    def determine_request_protocol(
        service_model: ServiceModel, content_type: str | None = None
    ) -> str:
        protocol = str(service_model.protocol)
        supported_protocols = service_model.metadata.get("protocols", [protocol])
        content_type = content_type if content_type is not None else ""
        if content_type in JSON_TYPES:
            protocol = "rest-json" if content_type == APPLICATION_JSON else "json"
        elif content_type.startswith("application/x-www-form-urlencoded"):
            protocol = "ec2" if "ec2" in supported_protocols else "query"
        if protocol not in supported_protocols:
>           raise NotImplementedError(
                f"Unsupported protocol [{protocol}] for service {service_model.service_name}"
            )
E           NotImplementedError: Unsupported protocol [rest-json] for service s3

.nox/pytest-3-14/lib/python3.14/site-packages/moto/core/request.py:50: NotImplementedError
```

</p>
</details>

--- a/moto/core/request.py
+++ b/moto/core/request.py
@@ -40,6 +40,10 @@ def determine_request_protocol(
     service_model: ServiceModel, content_type: str | None = None
 ) -> str:
     protocol = str(service_model.protocol)
+    # Short circuit protocol detection for S3 because the ContentType header
+    # is often set based on the MIME type of the object data being uploaded.
+    if service_model.service_name == "s3":
+        return protocol
     supported_protocols = service_model.metadata.get("protocols", [protocol])
     content_type = content_type if content_type is not None else ""
     if content_type in JSON_TYPES:

--- a/tests/test_s3/test_s3_multipart.py
+++ b/tests/test_s3/test_s3_multipart.py
@@ -1468,3 +1468,20 @@ def test_multipart_response_contains_checksum_attributes():
     (upload,) = bucket.multipart_uploads.all()
     assert upload.checksum_algorithm == "SHA256"
     assert upload.checksum_type == "COMPOSITE"
+
+
+@s3_aws_verified
+@pytest.mark.aws_verified()
+def test_create_multipart_upload_with_content_type(bucket_name=None):
+    # Regression test for https://github.com/getmoto/moto/issues/10010
+    s3_client = boto3.client("s3", "us-east-1")
+    s3_client.create_bucket(Bucket=bucket_name)
+    key = "my-key"
+    resp = s3_client.create_multipart_upload(
+        Bucket=bucket_name,
+        Key=key,
+        ContentType="application/json",
+    )
+    assert resp["Bucket"] == bucket_name
+    assert resp["Key"] == key
+    assert resp["UploadId"]

